plugins/out_stackdriver/stackdriver.c

```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
     return ctx->o->access_token;
 }
 
+/*
+* parse_labels
+* - Iterate throught the original payload (obj) and find out the entry that matches the labels_key 
+* - Convert all labels under labels_key to root-level `labels` field
+*/
+static msgpack_object *parse_labels(struct flb_stackdriver *ctx, msgpack_object *obj)
+{
+    int i;
+    int len;
+    msgpack_object_kv *kv = NULL;
+
+    if (!obj || obj->type != MSGPACK_OBJECT_MAP) {
+        return NULL;
+    }
+
+    len = flb_sds_len(ctx->labels_key);
+    for (i = 0; i < obj->via.map.size; i++) {
+        kv = &obj->via.map.ptr[i];
+        if (kv->key.via.str.size == len &&
+            strncasecmp(kv->key.via.str.ptr, ctx->labels_key, len) == 0) {
+            if (kv->val.type != MSGPACK_OBJECT_MAP) {
+                flb_plg_warn(ctx->ins, "the type of labels should be map");
+            }
+            return &kv->val;
+        }
+    }
+
+    flb_plg_debug(ctx->ins, "labels_key [%s] not found in the payload", 
+                  ctx->labels_key);
+    return NULL;
+}
+
+/* using packer mp_pck to pack obj except some special elements */
+static int pack_object_except(msgpack_object *obj, msgpack_packer mp_pck, 
+                              struct flb_stackdriver *ctx)
+{
+    int i, j;
+    int map_size;
+    int new_map_size;
+    int key_size;
+    int len;
+    int len_to_be_removed;
+    int key_not_found;
+    flb_sds_t removed;
+    msgpack_object_kv *kv = NULL;
+
+    /*
+     * array of elements that need to be removed from payload 
+     * more elements are required to be added here: labels, operations ...
+     */
+    flb_sds_t to_be_removed[] = 
+    {
+        ctx->labels_key
+    };
+
+    if (!obj) {
+        return -1;
+    }
+
+    map_size = obj->via.map.size;
+    new_map_size = map_size;
+    len_to_be_removed = sizeof(to_be_removed) / sizeof(to_be_removed[0]);
+    for (i = 0; i < map_size; i++) {
+        kv = &obj->via.map.ptr[i];
+        len = kv->key.via.str.size;
+        for (j = 0; j < len_to_be_removed; j++) {
+            removed = to_be_removed[j];
+            /* 
+             * check length of key to avoid partial matching
+             * e.g. labels key = labels && kv->key = labelss
+             */
+            if (flb_sds_len(removed) == len &&
+                strncasecmp(kv->key.via.str.ptr, removed, len) == 0) {
+                new_map_size -= 1;
+                break;
+            }
+        }
+    }
+
+    /* pack the original obj */ 
+    if (new_map_size == map_size) {
+        msgpack_pack_object(&mp_pck, *obj);
+        return 0;
+    }
+
+    msgpack_pack_map(&mp_pck, new_map_size);
+
+    for (i = 0; i < map_size; ++i) {
```

Comment:
This block is almost entirely identical to the block starting on line 342, though I'm not sure if we can improve it much. Still, it would be worth taking a look at filter_modifier.c, which does something almost identical to this. They seem to use a bool_map to track each entry and whether it should be included/excluded which makes the subsequent iteration to create the resulting map a bit more direct.

Another approach would be to move this to a helper function which takes an additional function pointer (action_when_key_not_found) which would be a simple counter in the first case, and some msg_pack functions in the second. This might be a little overkill though.

Feel free to leave as is; no action required.

Reply:
thanks